src/webgpu/shader/execution/expression/call/builtin/atomics/atomicCompareExchangeWeak.spec.ts

```diff
@@ -34,6 +177,337 @@ struct __atomic_compare_exchange_result<T> {
+    t.expectGPUBufferValuesEqual(outputBuffer, outputExpected);
+
+    // Read back exchanged buffer
+    const exchangedBufferResult = await t.readGPUBufferRangeTyped(exchangedBuffer, {
+      type: arrayType,
+      typedLength: exchangedBuffer.size / arrayType.BYTES_PER_ELEMENT,
+    });
+
+    // And the wg copy buffer should have been modified to a computed value for every third value,
+    // unless the comparison spuriously failed.
+    const wgCopyBufferExpected = new (typedArrayCtor(t.params.scalarType))(
+      wgNumElements * dispatchSize
+    );
+    wgCopyBufferExpected.forEach((_, i) => {
+      if (i % 3 === 0 && exchangedBufferResult.data[i]) {
+        wgCopyBufferExpected[i] = mapId.f(i * 2, numInvocations);
+      } else {
+        wgCopyBufferExpected[i] = i; // No change
+      }
+    });
+    t.expectGPUBufferValuesEqual(wgCopyBuffer, wgCopyBufferExpected);
+  });
+
+g.test('compare_exchange_weak_storage_advanced')
+  .specURL('https://www.w3.org/TR/WGSL/#atomic-rmw')
+  .desc(
+    `
+AS is storage or workgroup
+T is i32 or u32
+
+fn atomicCompareExchangeWeak(atomic_ptr: ptr<AS, atomic<T>, read_write>, cmp: T, v: T) -> __atomic_compare_exchange_result<T>
+
+struct __atomic_compare_exchange_result<T> {
+  old_value : T,    // old value stored in the atomic
+  exchanged : bool, // true if the exchange was done
+}
+`
+  )
+  .params(u =>
+    u
+      .combine('workgroupSize', workgroupSizes)
+      .combine('dispatchSize', dispatchSizes)
+      .combine('scalarType', ['u32', 'i32'])
+  )
+  .fn(async t => {
+    const numInvocations = t.params.workgroupSize * t.params.dispatchSize;
+    const scalarType = t.params.scalarType;
+
+    // Number of times each workgroup attempts to write the same value to the same memory address
+    const numWrites = 4;
+
+    const bufferNumElements = numInvocations * numWrites;
+    const defaultValue = 99999999;
+
+    const wgsl = `
+      @group(0) @binding(0)
+      var<storage, read_write> input : array<atomic<${scalarType}>>;
+
+      @group(0) @binding(1)
+      var<storage, read_write> output : array<${scalarType}>;
+
+      @group(0) @binding(2)
+      var<storage, read_write> exchanged : array<${scalarType}>;
+
+      @compute @workgroup_size(${t.params.workgroupSize})
+      fn main(
+          @builtin(global_invocation_id) global_invocation_id : vec3<u32>,
+          ) {
+        let id = ${scalarType}(global_invocation_id[0]);
+
+        // Each invocation attempts to write 'compare + 1' if 'compare' is at input[0], once per loop.
+        // Only one invocation per loop iteration should succeed. Note the workgroupBarrier() used
+        // to synchronize each invocation in the loop.
+
+        var compare = ${scalarType}(0);
+        for (var i = 0u; i < ${numWrites}u; i++) {
+          // Only one invocation will succeed at writing its value,
+          // or none if spurious failures occur.
+          let r = atomicCompareExchangeWeak(&input[0], compare, compare + 1);
```

Comment:
We discussed bouncing between two comparison values, so that a single spurious failure doesn't break the chain, and instead the passes after a spurious failure can resynchronise. This seems to be incrementing?

Reply:
Done.